File: lua/PerfTest.lua
-- Performance Test on Lua

-- Test 1: Execution time test
function test1()
    local StartTime = os.clock()

    for i = 1, 100000 do
        for j = 1, 100 do
            local a = 5 * 5;
            local b = a + 5;
        end

        for j = 1, 100 do
            local a = 5.5 * 5.5;
            local b = a + 5.5;
        end

        local array = {};
        for j = 1, 100 do
            array[j] = 0;
        end
        for j = 1, 100 do
            array[50] = array[70];
        end
        for j = 1, 100 do
            local a = "hello";
            local b = a .. "world";
        end
    end

    local EndTime = os.clock()

    local ExeTime = EndTime - StartTime

    -- print("Test 1 eplapsed time: " .. ExeTime)
    return ExeTime
end

-- Test 2: Execution time when calling C Function
function test2()
    local StartTime = os.clock()

    -- Load clib.dll
    require("clib")

    for i = 1, 100000000 do
        local a = sum_written_in_c(1, 2, 3)
    end

    local EndTime = os.clock()

    local ExeTime = EndTime - StartTime

    print("End test 2")
    -- print("Test 2 eplapsed time: " .. ExeTime)
    return ExeTime
end

-- Test 3: Memory test
function test3()
    for i = 1, 2 do
        local ints = {}
        for j = 1, 100000 do
            ints[j] = j
        end

        local floats = {}
        for j = 1, 100000 do
            floats[j] = j
        end

        local strings = {}
        for j = 1, 100000 do
            strings[j] = j
        end
    end
end

-- Run test here

-- Run test 3
-- test3()


IteratorTime = 10

Reduce = function (n, list)
    for key, value in pairs(list) do
        n = n + value
    end
    return n
end

-- Run test 1
-- test1()
TimeList1 = {}

for i = 1, IteratorTime do
    TimeList1[i] = test1()
end

ElapsedTime1 = Reduce(0, TimeList1) / IteratorTime

print("Test 1 elapsed time: " .. ElapsedTime1 .. "s")

-- Run test 2
test2()

TimeList2 = {}

for i = 1, IteratorTime do
    TimeList2[i] = test2()
end

ElapsedTime2 = Reduce(0, TimeList2) / IteratorTime

print("Test 2 elapsed time: " .. ElapsedTime2 .. "s")

File: cpp/PerfTest/PerfTest/PerfTest.cpp
// PerfTest.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include "pch.h"
#include <iostream>

#include <windows.h>

#include <stdio.h>
#include <string>
#include <vector>

#include <chrono>
#include <ctime>

#include <profileapi.h>

// Import DLL C Func here
//typedef int(__cdecl *MYPROC)();
typedef int(__cdecl *MYPROC)(double, double, double);

static double reduce(double* n, std::vector<double> list)
{
    for (int i = 0; i < list.size(); i++)
    {
        *n += list.at(i);
    }
    return *n;
}

static double test1()
{
    LARGE_INTEGER frequency;        // ticks per second
    LARGE_INTEGER t1, t2;           // ticks
    //double elapsedTime;

    // get ticks per second
    QueryPerformanceFrequency(&frequency);

    // start timer
    QueryPerformanceCounter(&t1);

    for (int i = 0; i < 100000; ++i)
    {
        for (int j = 0; j < 100; ++j)
        {
            int a = 5 * 5;
            int b = a + 5;
        }
        for (int j = 0; j < 100; ++j)
        {
            double a = 5.5 * 5.5;
            double b = a + 5.5;
        }
        int array[100];
        for (int j = 0; j < 100; ++j)
        {
            array[j] = 0;
        }
        for (int j = 0; j < 100; ++j)
        {
            array[50] = array[70];
        }
        for (int j = 0; j < 100; ++j)
        {
            std::string a = "hello";
            std::string b = a + "world";
        }
    }

    // stop timer
    QueryPerformanceCounter(&t2);

    // compute and print the elapsed time in millisec
    //elapsedTime = (t2.QuadPart - t1.QuadPart) * 1.0 / frequency.QuadPart;
    //std::cout << elapsedTime;

    return (t2.QuadPart - t1.QuadPart) * 1.0 / frequency.QuadPart;;
}

static double test2()
{
    LARGE_INTEGER frequency;        // ticks per second
    LARGE_INTEGER t1, t2;           // ticks

    // get ticks per second
    QueryPerformanceFrequency(&frequency);

    // start timer
    QueryPerformanceCounter(&t1);

    HINSTANCE hinstLib;
    MYPROC ProcSum;
    BOOL fFreeResult, fRunTimeLinkSuccess = FALSE;

    // Get a handle to the DLL module.
    hinstLib = LoadLibrary(TEXT("CLib.dll"));

    if (hinstLib != NULL) // If the handle is valid, try to get the function address.
    {
        ProcSum = (MYPROC)GetProcAddress(hinstLib, "sum_written_in_c");

        if (NULL != ProcSum) // If the function address is valid, call the function.
        {
            fRunTimeLinkSuccess = TRUE;
            for (int i = 0; i < 100000000; ++i)
                int a = (ProcSum)(1, 2, 3);
        }
        

        fFreeResult = FreeLibrary(hinstLib); // Free the DLL module.
    }

    if (!fRunTimeLinkSuccess) // If unable to call the DLL function, use an alternative.
        printf("Fail Test 2: Not loading CLib.dll");

    // stop timer
    QueryPerformanceCounter(&t2);

    return (t2.QuadPart - t1.QuadPart) * 1.0 / frequency.QuadPart;
}

static void test3()
{
    for (int i = 0; i < 2; ++i)
    {
        std::vector<int> ints;
        for (int j = 1; j < 100000; ++j)
        {
            ints.push_back(1);
        }
        std::vector<float> floats;
        for (int j = 1; j < 100000; ++j)
        {
            floats.push_back(1.1);
        }
        std::vector<std::string> strings;
        for (int j = 1; j < 100000; ++j)
        {
            strings.push_back("hello world");
        }
    }

    //return 0;
}

int main()
{
    //double* elapsedTime = 0;
    const int iteratorTime = 100;

    // Test 1

    std::vector<double> timeList1;

    for (int i = 0; i < iteratorTime; i++)
    {
        timeList1.push_back(test1());
    }

    //double test1Result = 0.0;
    double elapsedTime1Total = 0.0;

    reduce(&elapsedTime1Total, timeList1);

    std::cout << "Test 1 elapsed time: " << elapsedTime1Total / (double) iteratorTime << "s\n";    

    // Test 2

    //test2();

    std::vector<double> timeList2;

    for (int i = 0; i < iteratorTime; i++)
    {
        timeList2.push_back(test2());
    }

    //double test1Result = 0.0;
    double elapsedTime2Total = 0.0;

    reduce(&elapsedTime2Total, timeList2);

    std::cout << "Test 2 elapsed time: " << elapsedTime2Total / (double) iteratorTime << "s\n";

    return 0;
}
